Adding pagination to list codespaces

geramirez opened the pull request:
Fixes https://github.com/cli/cli/issues/4441

- Added pagination to for the `user/codespaces` endpoint. 
- depends on getting an empty list to stop pagination

geramirez committed:
Adding pagination to list codespaces

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

geramirez committed:
Fixing Has() issue due to go version

geramirez committed:
Merge remote-tracking branch 'upstream/trunk' into geramirez/codespaces-pagination

geramirez committed:
Add strconv

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
-	}
+	page := 1
+	per_page := 50
+	codespaces := []*Codespace{}
```
If you name the return variable you don't need this declaration

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
-		return nil, fmt.Errorf("error creating request: %w", err)
-	}
+	page := 1
+	per_page := 50
```
This per page has an inherent limit on the server-side right? If I pass 100, the server will ignore it and use 50?

geramirez replied:
right now, it will use the default if the per_page is out of range. I'm going to create another PR that uses the max if we request more along with a couple other updates.

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
-		return nil, fmt.Errorf("error making request: %w", err)
-	}
-	defer resp.Body.Close()
+	for {
```
```suggestion
	for page := 1;; page++ {
```

And you can remove the declaration above

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
+			page++
+		} else {
+			break
+		}
```
Switch the logic here to avoid the else branch and make the special condition the breaking condition.

```suggestion
		if len(response.Codespaces) == 0 {
		        break
		}
	        
	        codespaces = append(codespaces, response.Codespaces...)
```

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
+			return nil, jsonErrorResponse(b)
+		}
+
+		var response struct {
+			Codespaces []*Codespace `json:"codespaces"`
+		}
```
Does the server return the total? Should we capture that and use that as our breaking condition instead? It feels like the current logic will do an extra round-trip to get the empty list of codespaces when we can perhaps deduce that we don't need to call again from the total count.

geramirez replied:
Yes, we do have total count, I added it to the fake api in the test. Let me update the condition, I'll have to update the tests to catch mid-flight additions or deletions.

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -188,36 +189,54 @@ type CodespaceEnvironmentConnection struct {
+		if err != nil {
+			return nil, fmt.Errorf("error making request: %w", err)
+		}
+		defer resp.Body.Close()
```
Defer in loops are usually bad since they queue up and don't release resources until the whole function is done. Perhaps it's better to create a new inline function. Or a new method.

ie

```
fetchCodespaces := func() ([]*Codespace, error) {
  ...
  defer resp.BodyClose()
}

...

for {
   codespaces, err := fetchCodespaces()
}
```

geramirez committed:
Update internal/codespaces/api/api.go

Co-authored-by: Jose Garcia <josebalius@github.com>

geramirez committed:
Update internal/codespaces/api/api.go

Co-authored-by: Jose Garcia <josebalius@github.com>

geramirez committed:
Updated PR with suggestions

- created a new method  to avoid defer pileup issue
- removed extra api call at the end of the sequence

geramirez committed:
Adding additional tests for mid-flight deletions and additions

geramirez committed:
Adding a second condition just in case

geramirez commented on internal/codespaces/api/api.go:
```diff
@@ -186,16 +187,42 @@ type CodespaceEnvironmentConnection struct {
+		if err != nil {
+			return nil, fmt.Errorf("%w", err)
+		}
+		if len(codespaces) >= response.TotalCount || len(response.Codespaces) == 0 {
```
@josebalius, we don't need the second condition, but I added it as insurance, in case we ever (accidentally or on purpose) change the way totals work today.

josebalius replied:
Usually the way I do this with pagination is something like this:
```
if page * per_page >= response.TotalCount // we've reached the end
```

So if the total count is 101, at 50 per page, we have 3 pages, on page 3, it would `3 * 50 = 150` which is greater than 101. This removes the need to check the actual codespaces we've fetched.

mislav replied:
The easiest way is to use pagination headers as returned from the server. Example: https://github.com/cli/cli/blob/3652307cf3c6a8a48711e7c28a7b8cf6c9dcb1a3/pkg/cmd/api/pagination.go#L13-L23

So if there is a "next" link, follow it; otherwise pagination is done.

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -186,16 +187,42 @@ type CodespaceEnvironmentConnection struct {
 	HostPublicKeys []string `json:"hostPublicKeys"`
 }
 
+// codespacesListResponse is the request body for the `/user/codespaces` endpoint
```
```suggestion
// codespacesListResponse is the response body for the `/user/codespaces` endpoint
```

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -186,16 +187,42 @@ type CodespaceEnvironmentConnection struct {
+
 // ListCodespaces returns a list of codespaces for the user.
-func (a *API) ListCodespaces(ctx context.Context) ([]*Codespace, error) {
+func (a *API) ListCodespaces(ctx context.Context) (codespaces []*Codespace, err error) {
```
Can we update the docs for this method?

```
// ListCodespaces returns a list of codespaces for the user. 
// It consumes all pages returned by the API until all codespaces have been fetched.
```

josebalius commented on internal/codespaces/api/api.go:
```diff
@@ -186,16 +187,42 @@ type CodespaceEnvironmentConnection struct {
+		if len(codespaces) >= response.TotalCount || len(response.Codespaces) == 0 {
+			break
+		}
+		codespaces = append(codespaces, response.Codespaces...)
```
I know I recommended making break the "special" case. But does this need to come before the break condition now?
```
codespaces = append(codespaces, response.Codespaces...)

if ... {
  break
}
```

geramirez replied:
yeah, I actually didn't remove the extra call, whoops. I'll update the test to ensure that it's the case.

josebalius reviewed (commented):
Looking good! Left a few more comments, I think it's almost ready.

geramirez committed:
Update internal/codespaces/api/api.go

Co-authored-by: Jose Garcia <josebalius@github.com>

mislav commented on internal/codespaces/api/api.go:
```diff
@@ -186,16 +187,42 @@ type CodespaceEnvironmentConnection struct {
+	for page := 1; ; page++ {
+		response, err := a.fetchCodespaces(ctx, page)
+		if err != nil {
+			return nil, fmt.Errorf("%w", err)
```
There is no reason to wrap the error object here

mislav commented on internal/codespaces/api/api.go:
```diff
@@ -186,16 +187,42 @@ type CodespaceEnvironmentConnection struct {
+}
+
+func (a *API) fetchCodespaces(ctx context.Context, page int) (response *getCodespacesListResponse, err error) {
+	per_page := 50
```
Why not make the default per_page=100? Even though there can only be max 50 Codespaces right now, that number could increase in the future, and usually the maximum per_page number across other APIs is 100 anyway.

geramirez committed:
Updating docs and interation exit condition to not check the final page

geramirez committed:
Merge branch 'geramirez/codespaces-pagination' of https://github.com/geramirez/cli into geramirez/codespaces-pagination

mislav reviewed (changes requested):
@geramirez Thanks for working on this!

I think that `gh cs list` could be consistent with the functionality of the rest of GitHub CLI `list` commands: they all support a `-L/--limit` flag for the user to choose the number of items to fetch and display. Usually, the default limit is 30.

If the user passes a `--limit` that is over 100, pagination is performed until enough items were fetched to satisfy the number. Although, it sounds like codespaces will never have to perform pagination since one can have at most 50 codespaces for now. But it's still good to implement generic API pagination under the hood like you've been doing so that all this is future-proof.

geramirez committed:
Fetch 100 codespaces by default

geramirez commented:
@mislav I'm fine with updating the `gh cs list` command with the --limit parameter. Do you mind if I open an issue for it since I have to prioritize a couple other parts of the pagination work?  I'd be fine if someone else picks up the work too (it might be a good first issue for someone). 

Here are the details to make sure, let me know:

Issue Title:  Add support for `-L/--limit` parameter to the `gh cs list` command.
Acceptance criteria:
* default limit should be 30
*  using`-L/--limit` will cause pagination until the limit is reached or the are no more codespaces

mislav reviewed (approved):
Sure, logging this as follow-up item is fine! I will unblock my review then, since everything else seems fine in this PR 👍

josebalius reviewed (approved):
Thank you for doing this @geramirez!
